Rover/mode_circle.cpp

```diff
@@ -146,13 +149,17 @@ void ModeCircle::update()
 
     // Update distance to destination and distance to edge
     const Vector2f center_to_veh = curr_pos - config.center_pos;
-    _distance_to_destination = center_to_veh.length();
-    dist_to_edge_m = fabsf(_distance_to_destination - config.radius);
+    _distance_to_destination = (target.pos.tofloat() - curr_pos).length();
+    dist_to_edge_m = fabsf(center_to_veh.length() - config.radius);
 
     // Update depending on stage
     if (!reached_edge) {
         update_drive_to_radius();
-
+    } else if (dist_to_edge_m > config.radius * 0.2) {
+        // if more than 20% inside or outside circle radius, stop vehicle
+        GCS_SEND_TEXT(MAV_SEVERITY_ERROR, "Circle: vehicle unable to turn fast enough to track circle. Stopping vehicle.");
+        stop_vehicle();
+        return;
```

Comment:
We also come through here in loiter turns mission items. This is not such a great solution in that case. Also it looks like it will call send text at loop rate?

Reply:
Yeah, I should add in a flag to only send this message once...